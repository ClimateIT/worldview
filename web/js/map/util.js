import OlLayerTile from 'ol/layer/Tile';
import OlTileGridTileGrid from 'ol/tilegrid/TileGrid';
import OlSourceTileWMS from 'ol/source/TileWMS';
import util from '../util/util';

const ZOOM_DURATION = 250;

export function createWMSLayer(date, extent, layerId, res) {
  layerId = layerId.replace('_v6_NRT', '');
  layerId = layerId.replace('_v1_NRT', '');
  console.log(layerId);
  const parameters = {
    LAYERS: layerId,
    VERSION: '1.1.1',
  };
  console.log(date);
  const sourceOptions = {
    url: `https://gibs-{a-c}.earthdata.nasa.gov/wms/epsg4326/best/wms.cgi?TIME=${util.toISOStringSeconds(util.roundTimeOneMinute(date))}`,
    cacheSize: 4096,
    style: 'default',
    crossOrigin: 'anonymous',
    transition: 0,
    wrapX: true,
    params: parameters,
    tileGrid: new OlTileGridTileGrid({
      origin: [-180, 90],
      resolutions: res,
    }),
  };
  const layer = new OlLayerTile({
    preload: Infinity,
    extent,
    source: new OlSourceTileWMS(sourceOptions),
    minResolution: res[5],
  });
  layer.date = date;
  return layer;
}
/*
 * Setting a zoom action
 *
 * @function self.zoomAction
 * @static
 *
 * @param {Object} map - OpenLayers Map Object
 * @param {number} amount - Direction and
 *  amount to zoom
 * @param {number} duation - length of animation
 * @param {array} center - point to center zoom
 *
 * @returns {void}
 */
export function mapUtilZoomAction(map, amount, duration, center) {
  const zoomDuration = duration || ZOOM_DURATION;
  const centerPoint = center || undefined;
  const view = map.getView();
  const zoom = view.getZoom();
  view.animate({
    zoom: zoom + amount,
    duration: zoomDuration,
    center: centerPoint,
  });
}
export function getActiveLayerGroup(map, layerGroupString) {
  let group = null;
  const array = map.getLayers().getArray();
  for (let i = 0, len = array.length; i < len; i++) {
    const layerGroup = array[i];
    if (layerGroup.get('group') === layerGroupString) {
      group = layerGroup;
      break;
    }
  }
  return group;
}
