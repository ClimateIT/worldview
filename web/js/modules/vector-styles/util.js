import {
  assign as lodashAssign,
  find as lodashFind,
  get as lodashGet,
  groupBy as lodashGroupBy
} from 'lodash';
import { Stroke, Style, Fill, Circle } from 'ol/style';
import vectorDialog from '../../containers/vector-dialog';
import { setStyleFunction } from './selectors';
import { openCustomContent } from '../modal/actions';
import { selectVectorFeatures } from './actions';

export function getVectorStyleAttributeArray(layer) {
  var isCustomActive = false;
  var isMinActive = false;
  var isMaxActive = false;
  if (layer.custom) { isCustomActive = true; }
  if (layer.min) { isMinActive = true; }
  if (layer.max) { isMaxActive = true; }
  const styleObj = lodashAssign({}, { key: 'custom', value: layer.custom, isActive: isCustomActive });
  const minObj = lodashAssign({}, { key: 'min', value: layer.min, isActive: isMinActive });
  const maxObj = lodashAssign({}, { key: 'max', value: layer.max, isActive: isMaxActive });
  const attrArray = [];

  [styleObj, minObj, maxObj].forEach(obj => {
    if (obj.isActive) {
      attrArray.push({
        id: obj.key === 'custom' ? 'style' : obj.key,
        value: obj.value
      });
    } else {
      if (obj.isActive) {
        attrArray.push({
          id: obj.key === 'custom' ? 'style' : obj.key,
          value: ''
        });
      }
    }
  });
  return attrArray;
}

export function getMinValue(v) {
  return v.length ? v[0] : v;
}

export function getMaxValue(v) {
  return v.length ? v[v.length - 1] : v;
}
export function isConditional(item) {
  return Array.isArray(item) && item[0] === 'case';
}

export function selectedCircleStyle(style) {
  const styleImage = style.getImage();
  const fill = styleImage.getFill();
  const radius = styleImage.getRadius() * 1.25;
  return new Style({
    image: new Circle({
      radius: radius,
      stroke: new Stroke({
        color: 'white',
        width: 1
      }),
      fill: new Fill({
        color: fill.getColor().replace(/[^,]+(?=\))/, '0.5')
      })
    })
  });
}
export function selectedPolygonStyle(style) {
  const fill = style.getFill();
  const color = fill.getColor().replace(/[^,]+(?=\))/, '0.5')
  const stroke = style.getStroke();
  stroke.setColor('white');
  stroke.setWidth(0.5);
  fill.setColor(color);
  return style;
}
export function selectedStyleFunction(feature, styleArray) {
  if (styleArray.length !== 1) return styleArray;
  return styleArray.map((style) => {
    const type = feature.getType();
    switch (type) {
      case 'Point':
        return selectedCircleStyle(style);
      case 'Polygon':
        return selectedPolygonStyle(style);
      default:
        return style;
    }
  });
}
export function onMapClickGetVectorFeatures(e, map, store) {
  let metaArray = [];
  let selected = {};
  let defs = [];
  const state = store.getState();
  const lastSelection = state.vectorStyles.selected;
  const config = state.config;
  map.forEachFeatureAtPixel(e.pixel, function (feature, layer) {
    const def = lodashGet(layer, 'wv.def');
    if (!def) return;

    if (def.vectorData && def.vectorData.id && def.title) {
      const layerId = def.id;
      if (!selected[layerId]) selected[layerId] = [];
      let features = feature.getProperties();
      const vectorDataId = def.vectorData.id;
      const data = config.vectorData[vectorDataId];
      const properties = data.mvt_properties;
      const titleKey = lodashFind(properties, { Function: 'Identify' })['Identifier'];
      const title = features[titleKey];
      if (selected[layerId].includes(title)) return
      const obj = {
        legend: properties,
        features: features,
        id: vectorDataId,
        title: def.title || def.id,
        featureTitle: title
      };
      metaArray.push(obj);
      selected[layerId].push(title);
    }
  });
  if (Object.entries(selected).length || Object.entries(lastSelection).length) {
    store.dispatch(selectVectorFeatures(selected));
  }
  if (metaArray.length) {
    store.dispatch(openCustomContent('Vector-dialog' + e.pixel[0] + e.pixel[1],
      {
        backdrop: false,
        clickableBehindModal: true,
        desktopOnly: true,
        isDraggable: true,
        wrapClassName: 'vector-modal-wrap',
        modalClassName: 'vector-modal light',
        CompletelyCustomModal: vectorDialog,
        customProps: { vectorMetaObject: lodashGroupBy(metaArray, 'id') },
        onClose: () => {
          store.dispatch(selectVectorFeatures({}));
        }
      }
    ));
  };
}
export function updateVectorSelection(selectionObj, lastSelection, layers, type, state) {
  const vectorStyles = state.config.vectorStyles
  for (let [key, featureIdArray] of Object.entries(selectionObj)) {
    const def = lodashFind(layers, { id: key });
    setStyleFunction(def, def.vectorStyle.id, vectorStyles, null, state, { type, features: featureIdArray });
    if (lastSelection[key]) delete lastSelection[key];
  }
  for (let [key] of Object.entries(lastSelection)) {
    const def = lodashFind(layers, { id: key });
    setStyleFunction(
      def,
      def.vectorStyle.id,
      vectorStyles,
      null,
      state
    );
  }

}