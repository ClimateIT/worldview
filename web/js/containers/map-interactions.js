import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import OlCoordinates from '../components/map/ol-coordinates';
import vectorDialog from './vector-dialog';
import { onMapClickGetVectorFeatures } from '../modules/vector-styles/util';
import { openCustomContent, onClose } from '../modules/modal/actions';
import { selectVectorFeatures } from '../modules/vector-styles/actions';
import { groupBy as lodashGroupBy } from 'lodash';
import { changeCursor } from '../modules/map/actions';

export class MapInteractions extends React.Component {
  constructor(props) {
    super(props);
    this.mouseMove = this.mouseMove.bind(this);
    this.singleClick = this.singleClick.bind(this);
    this.registerMouseListeners();
  }

  registerMouseListeners() {
    this.props.mouseEvents.on('mousemove', this.mouseMove);
    this.props.mouseEvents.on('singleclick', this.singleClick);
  }

  singleClick(e, map) {
    const { lastSelected, openVectorDiaglog, onCloseModal, selectVectorFeatures, modalState, getDialogObject } = this.props;
    const isVectorModalOpen = modalState.id.includes('vector_dialog') && modalState.isOpen;
    const pixels = e.pixel;
    const clickObj = getDialogObject(pixels, map);
    const metaArray = clickObj.metaArray || [];
    const selected = clickObj.selected || {};
    const offsetLeft = clickObj.offsetLeft || 10;
    const offsetTop = clickObj.offsetTop || 100;
    const dialogId = isVectorModalOpen ? modalState.id : 'vector_dialog' + pixels[0] + pixels[1];

    if (metaArray.length) {
      openVectorDiaglog(dialogId, metaArray, offsetLeft, offsetTop);
    }
    if (Object.entries(selected).length || (Object.entries(lastSelected).length && !isVectorModalOpen)) {
      selectVectorFeatures(selected);
    } else if (isVectorModalOpen && !Object.entries(selected).length) {
      onCloseModal();
      selectVectorFeatures({});
    }
  }

  mouseMove(event, map, crs) {
    const pixels = map.getEventPixel(event);
    const coord = map.getCoordinateFromPixel(pixels);
    const { isShowingClick, changeCursor } = this.props;
    if (!coord) {
      this.clearCoord();
      return;
    }
    const hasFeatures = map.hasFeatureAtPixel(pixels);
    if (hasFeatures && !isShowingClick) {
      changeCursor(true);
    } else if (!hasFeatures && isShowingClick) {
      changeCursor(false);
    }
  }

  render() {
    const { isShowingClick, mouseEvents } = this.props;
    const mapClasses = isShowingClick ? 'wv-map' + ' cursor-pointer' : 'wv-map';

    return (
      <React.Fragment>
        <div id="wv-map" className={mapClasses} />
        <OlCoordinates mouseEvents={mouseEvents} />
      </React.Fragment>
    );
  }
}
const mapDispatchToProps = dispatch => ({
  selectVectorFeatures: (features) => {
    dispatch(selectVectorFeatures(features));
  },
  changeCursor: (bool) => {
    dispatch(changeCursor(bool));
  },
  onCloseModal: () => {
    dispatch(onClose());
  },
  openVectorDiaglog: (dialogId, metaArray, offsetLeft, offsetTop) => {
    dispatch(openCustomContent(dialogId,
      {
        backdrop: false,
        clickableBehindModal: true,
        desktopOnly: true,
        isDraggable: true,
        wrapClassName: 'vector-modal-wrap',
        modalClassName: 'vector-modal light',
        CompletelyCustomModal: vectorDialog,
        isResizable: true,
        vectorMetaObject: lodashGroupBy(metaArray, 'id'),
        width: 445,
        height: 300,
        offsetLeft,
        offsetTop,
        timeout: 150,
        onClose: () => {
          dispatch(selectVectorFeatures({}));
        }
      }
    ));
  }
});
function mapStateToProps(state) {
  return {
    modalState: state.modal,
    isShowingClick: state.map.isClickable,
    getDialogObject: (pixels, map) => { return onMapClickGetVectorFeatures(pixels, map, state); },
    lastSelected: state.vectorStyles.selected
  };
}
MapInteractions.propTypes = {
  changeCursor: PropTypes.func.isRequired,
  getDialogObject: PropTypes.func.isRequired,
  isShowingClick: PropTypes.bool.isRequired,
  modalState: PropTypes.object.isRequired,
  mouseEvents: PropTypes.object.isRequired,
  onCloseModal: PropTypes.func.isRequired,
  openVectorDiaglog: PropTypes.func.isRequired,
  selectVectorFeatures: PropTypes.func.isRequired,
  lastSelected: PropTypes.object
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MapInteractions);
