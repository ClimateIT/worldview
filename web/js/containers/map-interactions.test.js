import React from 'react';
import renderer from 'react-test-renderer';
import util from '../util/util';
import { OlCoordinates } from './map-interactions';
import { registerProjections } from '../fixtures';

let events;
let component;
let map;
let changeCursor;
let selectVectorFeatures;
let getDialogObject;
let openVectorDiaglog;
beforeEach(() => {
  registerProjections();
  events = util.events();
  changeCursor = jest.fn();
  openVectorDiaglog = jest.fn();
  selectVectorFeatures = jest.fn();
  getDialogObject = () => { return { metaArray: [0], selected: [1], offsetLeft: 100, offsetTop: 100 }; };
  component = renderer.create(
    <OlCoordinates
      mouseEvents={events}
      isShowingClick={false}
      changeCursor={changeCursor}
      getDialogObject={getDialogObject}
      openVectorDiaglog={openVectorDiaglog}
      selectVectorFeatures={selectVectorFeatures}
      lastSelected={{}}
      modalState={{ id: [], isOpen: false }}
    />
  );
  map = {
    getEventPixel: jest.fn(),
    getCoordinateFromPixel: () => [0, 0],
    hasFeatureAtPixel: () => false
  };
});

afterEach(() => {
  util.setCoordinateFormat('latlon-dd');
});

test('shows coordinates of (10, 20) when moving the mouse', () => {
  map.getCoordinateFromPixel = () => [10, 20];
  events.trigger('mousemove', {}, map, 'EPSG:4326');
  expect(component.toJSON()).toMatchSnapshot();
  expect(changeCursor.mock.calls.length).toBe(0);
});
test('shows 10°00.000\'N, 20°00.000\'E when set to degrees and minutes format', () => {
  util.setCoordinateFormat('latlon-dm');
  map.getCoordinateFromPixel = () => [10, 20];
  events.trigger('mousemove', {}, map, 'EPSG:4326');
  expect(component.toJSON()).toMatchSnapshot();
  expect(changeCursor.mock.calls.length).toBe(0);
});

test('clears coordinates when mouse moves off the map', () => {
  map.getCoordinateFromPixel = () => [10, 20];
  events.trigger('mousemove', {}, map, 'EPSG:4326');
  events.trigger('mouseout', {}, map, 'EPSG:4326');
  expect(component.toJSON()).toMatchSnapshot();
  expect(changeCursor.mock.calls.length).toBe(0);
});

test('reprojects (0,0) to (-45, 90) for EPSG:3413', () => {
  map.getCoordinateFromPixel = () => [0, 0];
  events.trigger('mousemove', {}, map, 'EPSG:3413');
  expect(component.toJSON()).toMatchSnapshot();
  expect(changeCursor.mock.calls.length).toBe(0);
});
test('if there is a feature at pixel dispatch changeCursor action', () => {
  map.hasFeatureAtPixel = () => true;
  events.trigger('mousemove', {}, map, 'EPSG:3413');
  expect(component.toJSON()).toMatchSnapshot();
  expect(changeCursor.mock.calls.length).toBe(1);
});
test('if there is a feature at pixel on click get dialod', () => {
  events.trigger('singleclick', { pixel: [0, 0] }, map, 'EPSG:4326');
  expect(component.toJSON()).toMatchSnapshot();
  expect(changeCursor.mock.calls.length).toBe(0);
  expect(selectVectorFeatures.mock.calls.length).toBe(1);
  expect(openVectorDiaglog.mock.calls.length).toBe(1);
});
