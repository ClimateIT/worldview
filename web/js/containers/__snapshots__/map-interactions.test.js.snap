// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check that cursor-hover class is not present 1`] = `
Array [
  <div
    className="wv-map"
    id="wv-map"
  />,
  <span
    id="vector-hover-label-tooltip"
    style={
      Object {
        "display": "none",
        "left": 0,
        "position": "absolute",
        "top": 0,
      }
    }
  >
     
  </span>,
]
`;
